Add support for RETURNING operator for ORCA

kooooootb opened the pull request:
Add support for RETURNING operator for Orca

If RETURNING is specified use Query's returningList to create the dxl:Ident
nodes and add them as child nodes of dxl:OutputColumns. Wrap dxl:LogicalInsert,
dxl:LogicalDelete and dxl:LogicalUpdate in the dxl:LogicalProject node that
contains dxl:ProjList, based on Query's returningList, if RETURNING has
expressions.

Change CLogicalInsert, CLogicalDelete, CLogicalUpdate, CLogicalDML, CPhysicalDML
so that they derive all columns from table that they are working with. Keep CTAS
behaviour unchanged.

Add dxl:ProjList to dxl:DMLInsert, dxl:DMLDelete and dxl:DMLUpdate as first
child node to store columns that will be received by parent node.

Store list of returning expressions in the targetlist field of the plan field of
the DML node. Add the targetListProj field to DML plan node and fill it with
list, which was previously stored in the targetlist field.

Change ExecInitDML to set up returning projection if DML plan node has non-null
targetlist. Change ExecDML to return tuples retrieved from ExecInsert and
ExecDelete if they are not null.

Update test samples in Orca unit tests and Greenplum regress tests.  Add tests
for INSERT, DELETE and UPDATE with RETURNING to Orca unit tests. Add
optimizer_trace_fallback=ON in "returning_gp" regress test. Split "returning_gp"
and "with" regress test expected files for planner and Orca.

kooooootb committed:
Add support for RETURNING operator for ORCA

This adds support for RETURNING operator for ORCA optimizer

returningList from query transforms into LogicalProject that wraps
around logical dml node in dxl query

dml nodes inside ORCA were changed to provide required columns for this
project node

nodeDML also was changed to return modified tuples to print them on
coordinator through projection

andr-sokolov commented:
bender build

andr-sokolov commented:
bender build

andr-sokolov commented:
@kooooootb, The tests detect errors.
Orca linter log:
```
docker run orca-linter:test
'/tmp/orca-clang-format.bVl/clang-format.yaml' -> 'src/backend/gporca/.clang-format'
src/backend/gporca/libgpopt/include/gpopt/operators/CPhysicalDML.h:80:22: error: code should be clang-formatted [-Wclang-format-violations]
```

Logs of regression tests with Orca:
[ORCA_x86-64_regression.diffs.tar.gz](https://github.com/user-attachments/files/18750362/ORCA_x86-64_regression.diffs.tar.gz)

kooooootb committed:
adapt leftover regress test to orca

kooooootb committed:
change expectance in regress with multiple DML CTE references

kooooootb committed:
also check CTEs when marking query as DML

kooooootb committed:
remove unused header

kooooootb committed:
fix tests

kooooootb committed:
fix double delete on logical delete

kooooootb committed:
fix gporca unit tests expected dumps

kooooootb committed:
revert comment edit

kooooootb committed:
gracer setting optimizer

kooooootb committed:
separate returning_gp regress expected to orca's and planner's, revert planner's to original

kooooootb committed:
add newline to returning_gp.out

andr-sokolov commented on src/test/regress/sql/with.sql:
```diff
@@ -1052,6 +1052,7 @@ select count(*) from rank_tbl where rank in (select rank from updated);
 DROP TABLE IF EXISTS t1;
 --end_ignore
 CREATE TABLE t1 (c1 int, c2 int) DISTRIBUTED RANDOMLY;
+SET optimizer = off;
```
Why is `optimizer` set to off here? Orca can build a plan for the query below.

kooooootb replied:
Got confused by comment above saying this test is for planner
I can divide expected file for orca and planner as i did for returning_gp

This test also has 2 other set optimizer=off, should i remove them too?

andr-sokolov replied:
As I see, other two tests don't relate to returning support, so let's don't change them. The expected file dividing is a good idea, because now we can test the query using Postres planner and Orca

kooooootb replied:
Reverted this, created separate expected file for ORCA

andr-sokolov commented on src/test/regress/expected/update_gp_optimizer.out:
```diff
@@ -258,19 +258,25 @@ WITH CTE AS (DELETE FROM t1 RETURNING *)
+                           ->  Hash
+                                 ->  Redistribute Motion 3:3  (slice2; segments: 3)
+                                       Hash Key: "outer".ColRef_0049
+                                       ->  Result
+                                             ->  Result
+                                                   ->  Delete
+                                                         ->  Result
```
Postgres planner does not add the Result nodes. Why are they added by Orca here and in other plans? Do we really need them?

kooooootb replied:
Removed mandatory LogicalProject (which converts to Result node) on top of DML node in query
This removed 2 upper Result s
Result s under DML nodes aren't seemed to be related to RETURNING, they are placed by some other DML logic
@andr-sokolov

kooooootb committed:
dont request extra result if no projection needed

kooooootb committed:
fix comment

kooooootb committed:
fix CTAS requesting motion for gathering output

kooooootb committed:
fix orca checking wrong target list for dml

kooooootb committed:
fix empty update output

kooooootb committed:
reorder proj lists in dml dxl plan node

output is first because most of other nodes expects it
fixed serializers, readers
Also small cleanup of unused changes

kooooootb committed:
check output existence by columns usage rather than by required columns

kooooootb committed:
Dont require distribution only for DML

SELECT on table with no columns is accepted but RETURNING cannot have
zero columns.
The only way to output from DML node is by using RETURNING.
So to require output for SELECT with zero columns tighten condition for
only DML operations

kooooootb committed:
fix column names in returning output

kooooootb committed:
linter fixes

kooooootb committed:
revert changind index of checked proj list in hash join

kooooootb committed:
adapt regress orca expected

kooooootb committed:
divide "with" regress test to planner and orca expectance files

kooooootb committed:
fix parsing after reordering

kooooootb committed:
actualize gporca test after reordering proj lists

kooooootb committed:
add gporca tests for DML with returning projection

kooooootb committed:
actualize gporca test after reordering proj lists

whitehawk commented on src/backend/gporca/libgpopt/src/operators/CLogicalDML.cpp:
```diff
@@ -386,6 +411,38 @@ CLogicalDML::OsPrint(IOstream &os) const
 		m_pcrSegmentId->OsPrint(os);
 	}
 
+	os << ", Output Columns: [";
```
Changes in 
 - CLogicalDML::OsPrint(IOstream &os)
 - CLogicalDelete::OsPrint(IOstream &os)
 - CLogicalInsert::OsPrint(IOstream &os)
 - CLogicalUpdate::OsPrint(IOstream &os)
 - CPhysicalDML::OsPrint(IOstream &os)
 
are duplicated.

And also much other newly added code in these classes is similar as well.

Is it possible to move such functionality into a parent class, from which these classes will be inherited?
Now, for example, `CLogicalDML` is inherited from `CLogical`.
It can be changed to: `CLogicalDML` is inherited from `CLogicalReturning`, which is inherited from `CLogical`. And `CLogicalReturning` will contain the code which is currenlty duplicated.

What do you think?

kooooootb replied:
I don't think that creating another class to inherit from gonna work out. I think these classes are too different to separate common logic in them

Most of my changes here include:
- Handling output columns in methods like HashValue() or Matches().
Duplicated lines could be gathered in methods in CLogicalReturning, which would be called in current methods, but i think this will take more space than it is taking now
- Adding ctor for use in PopCopyWithRemappedColumns() method in Logical classes
I can gather common logic in each paired ctors in some method, but it is quite different for e.g. CLogicalInsert and CLogicalDelete, as they store different states. This common logic is better to be placed near each of 4 classes, i can do it
- Extending OsPrint with output columns
Printing output columns really is duplicated but i think making another class to inherit from for extending debug print is overkill, i can create printing function somewhere in utils and call it here

whitehawk replied:
Please check quick&dirty patch [test-2025-03-24-1.patch](https://github.com/user-attachments/files/19414206/test-2025-03-24-1.patch). I still think that we could leverage from moving at least part of common functionality into the super class. Maybe you can find out even more parts to move.

kooooootb replied:
Applied it

whitehawk commented on src/test/regress/sql/returning_gp.sql:
```diff
@@ -31,7 +31,7 @@ returning distkey, partkey, t;
 update returning_parttab set partkey = 19 where partkey = 13 returning *;
 
--- update that moves the tuple across partitions (not supported)
+-- update that moves the tuple across partitions
```
I suggest adding 
```
SET optimizer_trace_fallback=ON;
```
to the beginning of the file to track if fallback to Postgres planner happens.

kooooootb replied:
done, could you please restart tests? 🙂
It traced only on insert with triggers scenario on my local try

whitehawk replied:
Sure, launched the tests [here](https://github.com/arenadata/gpdb/pull/1361). Regression passed fine.

whitehawk commented on src/backend/executor/nodeDML.c:
```diff
@@ -36,6 +36,28 @@ ExecDMLExplainEnd(PlanState *planstate, struct StringInfoData *buf)
+void
+RemapProjection(ProjectionInfo *projInfo, AttrMap *map)
+{
+	int *varNumbers = projInfo->pi_varNumbers;
```
Please use consistent indents. Currently there are tabs and spaces mixed here.

kooooootb replied:
fixed here, didnt find spaces in other files

kooooootb committed:
fix indent

kooooootb committed:
add tracing fallback to planner in returning_gp test

kooooootb committed:
Merge branch 'adb6dev' into orca_returning_dml_dev

whitehawk commented:
Following query fails with Orca:
```
create table t (a int) distributed by (a);
with cte as (insert into t  values (5) returning  a, gp_segment_id) select * from cte order by gp_segment_id;
```

It results in:
```
postgres=# create table t (a int) distributed by (a);
CREATE TABLE
postgres=# with cte as (insert into t  values (5) returning  a, gp_segment_id) select * from cte order by gp_segment_id;
ERROR:  Error on receive from seg0 slice1 127.0.1.1:6002 pid=266064: server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
```

Explain also doesn't work:
```
postgres=# explain with cte as (insert into t  values (5) returning  a, gp_segment_id) select * from cte order by gp_segment_id;
ERROR:  bogus varattno for OUTER_VAR var: 8 (ruleutils.c:6005)
```

While Postgres standard planner works fine:
```
postgres=# set optimizer = 0;
SET
postgres=# explain with cte as (insert into t  values (5) returning  a, gp_segment_id) select * from cte order by gp_segment_id;
                                 QUERY PLAN                                 
----------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.03..0.04 rows=1 width=0)
   Merge Key: t.gp_segment_id
   ->  Sort  (cost=0.03..0.04 rows=1 width=0)
         Sort Key: t.gp_segment_id
         ->  Insert on t  (cost=0.00..0.01 rows=1 width=0)
               ->  Result  (cost=0.00..0.01 rows=1 width=0)
 Optimizer: Postgres query optimizer
(7 rows)

postgres=# with cte as (insert into t  values (5) returning  a, gp_segment_id) select * from cte order by gp_segment_id;
 a | gp_segment_id 
---+---------------
 5 |             2
(1 row)
```

whitehawk commented on src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp:
```diff
@@ -4445,11 +4538,29 @@ CDXLNode *
+//		is responsible for freeing it.
+//---------------------------------------------------------------------------
+CDXLNode *
+CTranslatorQueryToDXL::TranslateExprToDXLProject(
```
Do we really need this function? It seems that it is called only from the old `TranslateExprToDXLProject()`

With the patch below I still can successfully build and run the code:
```diff
diff --git a/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp b/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp
index 9405432b5a0..f35f4a0b5b2 100644
--- a/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp
+++ b/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp
@@ -4538,27 +4538,11 @@ CDXLNode *
 CTranslatorQueryToDXL::TranslateExprToDXLProject(Expr *expr,
                                                                                                 const CHAR *alias_name,
                                                                                                 BOOL insist_new_colids)
-{
-       return TranslateExprToDXLProject(expr, alias_name, m_var_to_colid_map,
-                                                                        insist_new_colids);
-}
-
-//---------------------------------------------------------------------------
-//     @function:
-//             CTranslatorQueryToDXL::TranslateExprToDXLProject
-//
-//     @doc:
-//             Create a DXL project element node from the target list entry or var.
-//             The function allocates memory in the translator memory pool, and the caller
-//             is responsible for freeing it.
-//---------------------------------------------------------------------------
-CDXLNode *
-CTranslatorQueryToDXL::TranslateExprToDXLProject(
-       Expr *expr, const CHAR *alias_name, CMappingVarColId *var_to_colid_map,
-       BOOL insist_new_colids)
 {
        GPOS_ASSERT(NULL != expr);
 
+       CMappingVarColId *var_to_colid_map = m_var_to_colid_map;
+
        // construct a scalar operator
        CDXLNode *child_dxlnode = TranslateExprToDXL(expr, var_to_colid_map);
 
diff --git a/src/include/gpopt/translate/CTranslatorQueryToDXL.h b/src/include/gpopt/translate/CTranslatorQueryToDXL.h
index 0b5154bacf4..edf8e26db2a 100644
--- a/src/include/gpopt/translate/CTranslatorQueryToDXL.h
+++ b/src/include/gpopt/translate/CTranslatorQueryToDXL.h
@@ -302,11 +302,6 @@ private:
        CDXLNode *TranslateExprToDXLProject(Expr *expr, const CHAR *alias_name,
                                                                                BOOL insist_new_colids = false);
 
-       // translate a target list entry or a join alias entry into a project element using specified mapping
-       CDXLNode *TranslateExprToDXLProject(Expr *expr, const CHAR *alias_name,
-                                                                               CMappingVarColId *var_to_colid_map,
-                                                                               BOOL insist_new_colids = false);
-
        // translate a CTE into a DXL logical CTE operator
        CDXLNode *TranslateCTEToDXL(const RangeTblEntry *rte, ULONG rti,
                                                                ULONG current_query_level);
```

whitehawk replied:
In this case I think we can also get rid of `CDXLNode *TranslateExprToDXL(Expr *expr, CMappingVarColId *var_to_colid_map);` as well:


```diff
diff --git a/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp b/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp
index 9405432b5a0..c6aabc8cce9 100644
--- a/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp
+++ b/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp
@@ -3969,24 +3969,9 @@ CTranslatorQueryToDXL::TranslateDerivedTablesToDXL(const RangeTblEntry *rte,
 //---------------------------------------------------------------------------
 CDXLNode *
 CTranslatorQueryToDXL::TranslateExprToDXL(Expr *expr)
-{
-       return TranslateExprToDXL(expr, m_var_to_colid_map);
-}
-
-//---------------------------------------------------------------------------
-//     @function:
-//             CTranslatorQueryToDXL::TranslateExprToDXL
-//
-//     @doc:
-//             Translate the Expr into a CDXLScalar node with specified mapping
-//
-//---------------------------------------------------------------------------
-CDXLNode *
-CTranslatorQueryToDXL::TranslateExprToDXL(Expr *expr,
-                                                                                 CMappingVarColId *var_to_colid_map)
 {
        CDXLNode *scalar_dxlnode =
-               m_scalar_translator->TranslateScalarToDXL(expr, var_to_colid_map);
+               m_scalar_translator->TranslateScalarToDXL(expr, m_var_to_colid_map);
        GPOS_ASSERT(NULL != scalar_dxlnode);
 
        return scalar_dxlnode;
@@ -4538,29 +4523,11 @@ CDXLNode *
 CTranslatorQueryToDXL::TranslateExprToDXLProject(Expr *expr,
                                                                                                 const CHAR *alias_name,
                                                                                                 BOOL insist_new_colids)
-{
-       return TranslateExprToDXLProject(expr, alias_name, m_var_to_colid_map,
-                                                                        insist_new_colids);
-}
-
-//---------------------------------------------------------------------------
-//     @function:
-//             CTranslatorQueryToDXL::TranslateExprToDXLProject
-//
-//     @doc:
-//             Create a DXL project element node from the target list entry or var.
-//             The function allocates memory in the translator memory pool, and the caller
-//             is responsible for freeing it.
-//---------------------------------------------------------------------------
-CDXLNode *
-CTranslatorQueryToDXL::TranslateExprToDXLProject(
-       Expr *expr, const CHAR *alias_name, CMappingVarColId *var_to_colid_map,
-       BOOL insist_new_colids)
 {
        GPOS_ASSERT(NULL != expr);
 
        // construct a scalar operator
-       CDXLNode *child_dxlnode = TranslateExprToDXL(expr, var_to_colid_map);
+       CDXLNode *child_dxlnode = TranslateExprToDXL(expr);
 
        // get the id and alias for the proj elem
        ULONG project_elem_id;
diff --git a/src/include/gpopt/translate/CTranslatorQueryToDXL.h b/src/include/gpopt/translate/CTranslatorQueryToDXL.h
index 0b5154bacf4..3355d510c67 100644
--- a/src/include/gpopt/translate/CTranslatorQueryToDXL.h
+++ b/src/include/gpopt/translate/CTranslatorQueryToDXL.h
@@ -211,10 +211,6 @@ private:
        // translate an Expr into CDXLNode
        CDXLNode *TranslateExprToDXL(Expr *expr);
 
-       // translate an Expr into CDXLNode with specified mapping
-       CDXLNode *TranslateExprToDXL(Expr *expr,
-                                                                CMappingVarColId *var_to_colid_map);
-
        // translate the JoinExpr (inside FromExpr) into a CDXLLogicalJoin node
        CDXLNode *TranslateJoinExprInFromToDXL(JoinExpr *join_expr);
 
@@ -302,11 +298,6 @@ private:
        CDXLNode *TranslateExprToDXLProject(Expr *expr, const CHAR *alias_name,
                                                                                BOOL insist_new_colids = false);
 
-       // translate a target list entry or a join alias entry into a project element using specified mapping
-       CDXLNode *TranslateExprToDXLProject(Expr *expr, const CHAR *alias_name,
-                                                                               CMappingVarColId *var_to_colid_map,
-                                                                               BOOL insist_new_colids = false);
-
        // translate a CTE into a DXL logical CTE operator
        CDXLNode *TranslateCTEToDXL(const RangeTblEntry *rte, ULONG rti,
                                                                ULONG current_query_level);

```

kooooootb replied:
removed them
i switched to shuffling m_var_colid_map in ProcessReturningList at some point and forgot them

whitehawk commented on src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp:
```diff
@@ -1214,7 +1278,14 @@ CTranslatorQueryToDXL::TranslateDeleteQueryToDXL()
+	CDXLNode *log_delete_dxlnode =
+		GPOS_NEW(m_mp) CDXLNode(m_mp, delete_dxlop, query_dxlnode);
+
+	CRefCount::SafeRelease(m_dxl_query_output_cols);
```
```
CRefCount::SafeRelease(m_dxl_query_output_cols);
```
is done each time before `ProcessReturningList()` is called. Can we move it inside `ProcessReturningList()` to reduce code duplication?

kooooootb replied:
moved, thanks, didnt notice that

kooooootb committed:
reduplicate deleting output cols in query2dxl

kooooootb committed:
remove unused methods

kooooootb committed:
some leftovers

kooooootb committed:
copy columns with their usage, this will reduced unused columns in ctes, only for optimization, not critical

kooooootb committed:
set result tuple slot to returning after exeecinit because sort node checks it before executing dml node

kooooootb committed:
store returning projection in DML plan node's targetlist, save previously stored targetlist in separate field

Parent nodes and explain logic is looking into plan node's target list to find out which tuples will be returned by node, so we store in DML targetlist targetentries which will be outputted
Currently targetlist stores projection for applying projections on child's results, so we store this targetlist in separate field and shuffle it with returning targetlist when needed

kooooootb committed:
adapt with_clause test

kooooootb committed:
add test case to with_clause - order by system column

kooooootb commented:
@whitehawk fixed issue with
create table t (a int) distributed by (a);
with cte as (insert into t  values (5) returning  a, gp_segment_id) select * from cte order by gp_segment_id;

Problem was that orderby node checked targetlist of DML node to get description of tuple which should be returned from it
DML node stored originally in its targetlist projections for its child's Result node to get values e.g. to insert
I switched targetlist to storing returning entries instead and stored targetlist for Results' in separate field in DML plan node which is used in DML's execution

kooooootb committed:
fix gporca unit test dumps after copying columns with their usage

kooooootb committed:
reindent new gporca minidumps

kooooootb committed:
fix output of DML nodes in plan in regress tests after fixing targetlist

whitehawk commented:
> with cte as (insert into t  values (5) returning  a, gp_segment_id) select * from cte order by gp_segment_id;

It doesn't fail now, but the plan look a bit weird:
```
postgres=# explain with cte as (insert into t  values (5) returning  a, gp_segment_id) select * from cte order by gp_segment_id;
                                  QUERY PLAN                                  
------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..0.01 rows=1 width=8)
   Merge Key: gp_segment_id
   ->  Sort  (cost=0.00..0.01 rows=1 width=8)
         Sort Key: gp_segment_id
         ->  Insert  (cost=0.00..0.01 rows=1 width=8)
               ->  Result  (cost=0.00..0.00 rows=1 width=8)
                     ->  Result  (cost=0.00..0.00 rows=1 width=4)
                           ->  Result  (cost=0.00..0.00 rows=1 width=4)
                                 ->  Result  (cost=0.00..0.00 rows=1 width=1)
 Optimizer: Pivotal Optimizer (GPORCA)
(10 rows)
```
as you can see, there are 4 `Result` nodes.
While standard Postgres planner provides almost the same, but with 1 `Result` node:
```
postgres=# set optimizer=0;
SET
postgres=# explain with cte as (insert into t  values (5) returning  a, gp_segment_id) select * from cte order by gp_segment_id;
                                 QUERY PLAN                                 
----------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.03..0.04 rows=1 width=0)
   Merge Key: t.gp_segment_id
   ->  Sort  (cost=0.03..0.04 rows=1 width=0)
         Sort Key: t.gp_segment_id
         ->  Insert on t  (cost=0.00..0.01 rows=1 width=0)
               ->  Result  (cost=0.00..0.01 rows=1 width=0)
 Optimizer: Postgres query optimizer
(7 rows)
``` 

Can the excessive `Result` nodes be removed?

whitehawk commented on src/include/gpopt/translate/CTranslatorDXLToPlStmt.h:
```diff
@@ -137,6 +137,12 @@ class CTranslatorDXLToPlStmt
 	// command type
 	CmdType m_cmd_type;
 
+	// is dml operation requiring returning
```
The comment sounds a bit bulky, I suggest to re-write it, smth like:
```suggestion
	// does dml operation require returning
```

kooooootb replied:
fixed it

kooooootb committed:
fix comment

kooooootb committed:
separate returning functionality in CLogicalReturning

kooooootb committed:
fixes CLogicalReturning

kooooootb committed:
fix linter

kooooootb commented:
@whitehawk could you pls restart tests, i guess only linter tests failed, i fixed them

whitehawk commented on src/backend/gporca/libgpopt/src/operators/CLogicalDelete.cpp:
```diff
@@ -108,10 +139,11 @@ CLogicalDelete::Matches(COperator *pop) const
 }
-
 //---------------------------------------------------------------------------
+
```
This new line is excessive.

kooooootb replied:
moved it up

whitehawk commented on src/backend/gporca/libgpopt/src/operators/CLogicalInsert.cpp:
```diff
@@ -106,6 +138,9 @@ CLogicalInsert::HashValue() const
 	ulHash =
 		gpos::CombineHashes(ulHash, CUtils::UlHashColArray(m_pdrgpcrSource));
 
+	ulHash =
```
I think we can move hash calculation for output columns into the base class. It will reduce the delta, and will be more future proof. I think it is more appropriate, if child class asks its base class for its hash value, instead of calculating it itself from the base class'es properties.
```diff
diff --git a/src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalReturning.h b/src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalReturning.h
index cfde0e0b14a..e041b74a081 100644
--- a/src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalReturning.h
+++ b/src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalReturning.h
@@ -66,6 +66,9 @@ public:
        // debug print
        virtual IOstream &OsPrint(IOstream &) const;
 
+       // hash function
+       virtual ULONG HashValue() const;
+
 };     // class CLogicalReturning
 
 }  // namespace gpopt
diff --git a/src/backend/gporca/libgpopt/src/operators/CLogicalDML.cpp b/src/backend/gporca/libgpopt/src/operators/CLogicalDML.cpp
index d05241e4e31..50ae2c94bcd 100644
--- a/src/backend/gporca/libgpopt/src/operators/CLogicalDML.cpp
+++ b/src/backend/gporca/libgpopt/src/operators/CLogicalDML.cpp
@@ -155,13 +155,11 @@ CLogicalDML::Matches(COperator *pop) const
 ULONG
 CLogicalDML::HashValue() const
 {
-       ULONG ulHash = gpos::CombineHashes(COperator::HashValue(),
+       ULONG ulHash = gpos::CombineHashes(CLogicalReturning::HashValue(),
                                                                           m_ptabdesc->MDId()->HashValue());
        ulHash = gpos::CombineHashes(ulHash, gpos::HashPtr<CColRef>(m_pcrAction));
        ulHash =
                gpos::CombineHashes(ulHash, CUtils::UlHashColArray(m_pdrgpcrSource));
-       ulHash =
-               gpos::CombineHashes(ulHash, CUtils::UlHashColArray(m_pdrgpcrOutput));
        ulHash = gpos::CombineHashes(ulHash, gpos::HashPtr<CColRef>(m_pcrTableOid));
 
        if (EdmlDelete == m_edmlop || EdmlUpdate == m_edmlop)
diff --git a/src/backend/gporca/libgpopt/src/operators/CLogicalDelete.cpp b/src/backend/gporca/libgpopt/src/operators/CLogicalDelete.cpp
index a2fe150033a..68a487994ac 100644
--- a/src/backend/gporca/libgpopt/src/operators/CLogicalDelete.cpp
+++ b/src/backend/gporca/libgpopt/src/operators/CLogicalDelete.cpp
@@ -154,11 +154,9 @@ CLogicalDelete::Matches(COperator *pop) const
 ULONG
 CLogicalDelete::HashValue() const
 {
-       ULONG ulHash = gpos::CombineHashes(COperator::HashValue(),
+       ULONG ulHash = gpos::CombineHashes(CLogicalReturning::HashValue(),
                                                                           m_ptabdesc->MDId()->HashValue());
        ulHash = gpos::CombineHashes(ulHash, CUtils::UlHashColArray(m_pdrgpcr));
-       ulHash =
-               gpos::CombineHashes(ulHash, CUtils::UlHashColArray(m_pdrgpcrOutput));
        ulHash = gpos::CombineHashes(ulHash, gpos::HashPtr<CColRef>(m_pcrCtid));
        ulHash =
                gpos::CombineHashes(ulHash, gpos::HashPtr<CColRef>(m_pcrSegmentId));
diff --git a/src/backend/gporca/libgpopt/src/operators/CLogicalInsert.cpp b/src/backend/gporca/libgpopt/src/operators/CLogicalInsert.cpp
index d46fe6cfcd1..5d44d54ea7f 100644
--- a/src/backend/gporca/libgpopt/src/operators/CLogicalInsert.cpp
+++ b/src/backend/gporca/libgpopt/src/operators/CLogicalInsert.cpp
@@ -133,14 +133,11 @@ CLogicalInsert::Matches(COperator *pop) const
 ULONG
 CLogicalInsert::HashValue() const
 {
-       ULONG ulHash = gpos::CombineHashes(COperator::HashValue(),
+       ULONG ulHash = gpos::CombineHashes(CLogicalReturning::HashValue(),
                                                                           m_ptabdesc->MDId()->HashValue());
        ulHash =
                gpos::CombineHashes(ulHash, CUtils::UlHashColArray(m_pdrgpcrSource));
 
-       ulHash =
-               gpos::CombineHashes(ulHash, CUtils::UlHashColArray(m_pdrgpcrOutput));
-
        return ulHash;
 }
 
diff --git a/src/backend/gporca/libgpopt/src/operators/CLogicalReturning.cpp b/src/backend/gporca/libgpopt/src/operators/CLogicalReturning.cpp
index b4e5b00ab1e..52c682b5c7b 100644
--- a/src/backend/gporca/libgpopt/src/operators/CLogicalReturning.cpp
+++ b/src/backend/gporca/libgpopt/src/operators/CLogicalReturning.cpp
@@ -120,4 +120,21 @@ CLogicalReturning::OsPrint(IOstream &os) const
        return os;
 }
 
+//---------------------------------------------------------------------------
+//     @function:
+//             CLogicalReturning::HashValue
+//
+//     @doc:
+//             Hash function
+//
+//---------------------------------------------------------------------------
+ULONG
+CLogicalReturning::HashValue() const
+{
+       ULONG ulHash = gpos::CombineHashes(COperator::HashValue(),
+                                                                          CUtils::UlHashColArray(m_pdrgpcrOutput));
+
+       return ulHash;
+}
+
 // EOF
diff --git a/src/backend/gporca/libgpopt/src/operators/CLogicalUpdate.cpp b/src/backend/gporca/libgpopt/src/operators/CLogicalUpdate.cpp
index 825831584b8..2e9ae870bc0 100644
--- a/src/backend/gporca/libgpopt/src/operators/CLogicalUpdate.cpp
+++ b/src/backend/gporca/libgpopt/src/operators/CLogicalUpdate.cpp
@@ -179,14 +179,12 @@ CLogicalUpdate::Matches(COperator *pop) const
 ULONG
 CLogicalUpdate::HashValue() const
 {
-       ULONG ulHash = gpos::CombineHashes(COperator::HashValue(),
+       ULONG ulHash = gpos::CombineHashes(CLogicalReturning::HashValue(),
                                                                           m_ptabdesc->MDId()->HashValue());
        ulHash =
                gpos::CombineHashes(ulHash, CUtils::UlHashColArray(m_pdrgpcrDelete));
        ulHash =
                gpos::CombineHashes(ulHash, CUtils::UlHashColArray(m_pdrgpcrInsert));
-       ulHash =
-               gpos::CombineHashes(ulHash, CUtils::UlHashColArray(m_pdrgpcrOutput));
        ulHash = gpos::CombineHashes(ulHash, gpos::HashPtr<CColRef>(m_pcrCtid));
        ulHash =
                gpos::CombineHashes(ulHash, gpos::HashPtr<CColRef>(m_pcrSegmentId));

```

kooooootb replied:
done

whitehawk commented on src/backend/gporca/libgpopt/src/operators/CLogicalInsert.cpp:
```diff
@@ -124,9 +159,24 @@ CLogicalInsert::PopCopyWithRemappedColumns(CMemoryPool *mp,
 	CColRefArray *colref_array =
 		CUtils::PdrgpcrRemap(mp, m_pdrgpcrSource, colref_mapping, must_exist);
+
+	CColRefArray *pdrgpcrOutput = NULL;
```
Can we move this part of code to the base class somehow? It is also duplicated in the child classes...

kooooootb replied:
created copy columns method in base class

whitehawk commented on src/backend/gporca/libgpopt/src/base/CUtils.cpp:
```diff
@@ -3276,6 +3277,54 @@ CUtils::FLogicalDML(COperator *pop)
+
+	while (exprsToCheck->Size() > 0)
+	{
+		CExpression *pexprCurrent = (*exprsToCheck)[exprsToCheck->Size() - 1];
+		exprsToCheck->RemoveLast();
```
`RemoveLast()` should 
```
// return the last element of the array and at the same time remove it from the array
```
so maybe
```suggestion
		CExpression *pexprCurrent = exprsToCheck->RemoveLast();
```
?

kooooootb replied:
done

whitehawk commented on src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalReturning.h:
```diff
@@ -0,0 +1,75 @@
+
+protected:
+	// table descriptor
+	CTableDescriptor *m_ptabdesc;
```
So, as `m_ptabdesc` was moved here. I think that the access method
 
```
	// return table's descriptor
	CTableDescriptor *
	Ptabdesc() const
	{
		return m_ptabdesc;
	}
```

can also be moved out from the child classes.

kooooootb replied:
done

whitehawk commented on src/backend/gporca/libgpopt/src/base/CUtils.cpp:
```diff
@@ -3276,6 +3277,54 @@ CUtils::FLogicalDML(COperator *pop)
+
+	CLogicalInsert *popInsert = CLogicalInsert::PopConvert(pop);
+
+	return IMDId::EmdidGPDBCtas == popInsert->Ptabdesc()->MDId()->MdidType();
```
what if `popInsert->Ptabdesc()` returns NULL? Or `MDId()`?

kooooootb replied:
You mean I should add some asserts for it?
There are some places where this chain is taken, check was taken from CTranslatorExprToDXL.cpp in PdxlnDML
	if (IMDId::EmdidGPDBCtas == popDML->Ptabdesc()->MDId()->MdidType())
	{
		return PdxlnCTAS(pexpr, pdrgpdsBaseTables, pulNonGatherMotions, pfDML);
	}

whitehawk replied:
I think adding asserts should be Ok.

kooooootb replied:
done

kooooootb committed:
move empty line

kooooootb committed:
FHasLogicalDML get last from RemoveLast

kooooootb committed:
descriptor getter to base class

kooooootb committed:
hash value and matches to base class

kooooootb committed:
forgot CLogicalDML

kooooootb committed:
deriveketcollection to base class

kooooootb committed:
add asserts for popInsert->Ptabdesc()->MDId()

kooooootb committed:
copy columns to base class

whitehawk commented on src/backend/gporca/libgpopt/src/base/CUtils.cpp:
```diff
@@ -3276,6 +3277,56 @@ CUtils::FLogicalDML(COperator *pop)
+		}
+
+		const ULONG arity = pexprCurrent->Arity();
+		for (ULONG ul = 0; ul < arity; ++ul)
```
```suggestion
		for (ULONG ul = 0; ul < pexprCurrent->Arity(); ++ul)
```

kooooootb replied:
done

whitehawk commented on src/backend/gporca/libgpopt/src/operators/CLogicalInsert.cpp:
```diff
@@ -86,7 +117,7 @@ CLogicalInsert::Matches(COperator *pop) const
 	CLogicalInsert *popInsert = CLogicalInsert::PopConvert(pop);
 
-	return m_ptabdesc->MDId()->Equals(popInsert->Ptabdesc()->MDId()) &&
+	return CLogicalReturning::MatchesReturning(popInsert) &&
```
Can we use usual 
```
	// match function
	virtual BOOL Matches(COperator *pop) const;
```
in the base class? Why do we need a separate `MatchesReturning()`?

kooooootb replied:
It has different CLogicalReturning in argument
If we rename it it will hide overloaded virtual function
And we can't check if COperator is derived from CLogicalReturning as Eopid is set in derived classes

whitehawk commented on src/backend/gporca/libgpopt/src/base/CQueryContext.cpp:
```diff
@@ -240,6 +239,28 @@ CQueryContext::PqcGenerate(CMemoryPool *mp, CExpression *pexpr,
+		CExpressionArray *pdrgpexpr = poptctxt->Pcteinfo()->PdrgPexpr(mp);
+
+		const ULONG size = pdrgpexpr->Size();
+		for (ULONG ul = 0; ul < size; ul++)
```
```suggestion
		for (ULONG ul = 0; ul < pdrgpexpr->Size(); ul++)
```

kooooootb replied:
done

whitehawk commented on src/backend/gporca/libgpopt/src/base/CQueryContext.cpp:
```diff
@@ -240,6 +239,28 @@ CQueryContext::PqcGenerate(CMemoryPool *mp, CExpression *pexpr,
+			}
+		}
+
+		pdrgpexpr->Release();
```
We didn't call `AddRef()`, so do we need to `Release()` here?

kooooootb replied:
PdrgPexpr creates new array of CTEs, each CTE is also AddRef'ed there

whitehawk commented on src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp:
```diff
@@ -1407,6 +1408,39 @@ CXformUtils::FTriggersExist(CLogicalDML::EDMLOperator edmlop,
+//
+//---------------------------------------------------------------------------
+BOOL
+CXformUtils::FTriggersExist(CLogicalDML::EDMLOperator edmlop,
```
This function duplicates a lot from
```
BOOL
CXformUtils::FTriggersExist(CLogicalDML::EDMLOperator edmlop,
							CTableDescriptor *ptabdesc, BOOL fBefore)
```
which is right above. Can we rework these 2 functions to avoid code duplication (by separatin the common part...)?

kooooootb replied:
separated common logic in FTriggersExistInner

whitehawk commented on src/backend/executor/nodeDML.c:
```diff
@@ -36,6 +36,28 @@ ExecDMLExplainEnd(PlanState *planstate, struct StringInfoData *buf)
+	int *varNumbers = projInfo->pi_varNumbers;
+	int numSimpleVars = projInfo->pi_numSimpleVars;
+
+	for (int i = 0; i < numSimpleVars;++i)
+	{
+		varNumbers[i] = attrMap(map, varNumbers[i]);
+	}
```
The surrounding code doesn't use braces for 1-line code blocks, so we should also follow this way to keep code style consistent (though personally I'd prefer to keep braces :) ).
```suggestion
	for (int i = 0; i < numSimpleVars; ++i)
		varNumbers[i] = attrMap(map, varNumbers[i]);
```

kooooootb replied:
done

whitehawk commented on src/backend/executor/nodeDML.c:
```diff
@@ -348,9 +447,32 @@ ExecEndDML(DMLState *node)
 	ExecFreeExprContext(&node->ps);
-	ExecClearTuple(node->ps.ps_ResultTupleSlot);
+	if (node->ps.ps_ResultTupleSlot != NULL)
+	{
```
braces are not used for 1-line code blocks throughout this file. We need to keep style consistent.

kooooootb replied:
done

kooooootb committed:
calc arity on place

kooooootb committed:
calc size in place

kooooootb committed:
nodeDML braces

kooooootb committed:
nodeDML braces

kooooootb committed:
common FTriggersExist logic in FTriggersExistInner

whitehawk commented on src/backend/executor/nodeDML.c:
```diff
@@ -45,158 +65,199 @@ ExecDMLExplainEnd(PlanState *planstate, struct StringInfoData *buf)
-			node->ps.state->es_result_partitions->part->parrelid &&
-			action != DML_DELETE)
-			makePartitionCheckMap(node->ps.state, relInfo);
+		/* restore returning result tuple and taregt list*/
```
```suggestion
		/* restore returning result tuple and target list*/
```

kooooootb replied:
done

whitehawk commented on src/backend/executor/nodeDML.c:
```diff
@@ -45,158 +65,199 @@ ExecDMLExplainEnd(PlanState *planstate, struct StringInfoData *buf)
+		Assert(outerNode != NULL);
 
-	Assert(outerNode != NULL);
+		/* Temporary restore result tuple slot for use in next projection */
```
Maybe
```suggestion
		/* Temporary store result tuple slot for use in next projection */
```
as we are storing old values for future use?

kooooootb replied:
I meant here we are restoring resultTuple to ps_ResultTupleSlot which will be used just in ExecProject on 112 line
On 118 line we are restoring resultTuple to ps_ResultTupleSlot which will be used returning projection

whitehawk commented:
```
SET optimizer_trace_fallback=ON;

CREATE TABLE returning_parttab (distkey int4, partkey int4, t text)
DISTRIBUTED BY (distkey)
PARTITION BY RANGE (partkey) (START (1) END (10));

explain with cte as (select max(gp_segment_id) from returning_parttab) insert into returning_parttab values (1, 5, 'test') returning *, 1 + (select * from cte);

```

results in 
```
postgres=# explain with cte as (select max(gp_segment_id) from returning_parttab) insert into returning_parttab values (1, 5, 'test') returning *, 1 + (select * from cte);
INFO:  GPORCA failed to produce a plan, falling back to planner
DETAIL:  CNormalizer.cpp:1467: Failed assertion: FLocalColsSubsetOfInputCols(mp, pexprOutput)
Stack trace:
1    0x000064d612321b5f gpos::CErrorContext::Record + 191
2    0x000064d612322d0e gpos::CException::Raise + 278
3    0x000064d61254b1c4 gpopt::CNormalizer::PexprPullUpProjections + 270
4    0x000064d61269e414 gpopt::CXformSubqueryUnnest::PexprSubqueryUnnest + 1096
5    0x000064d61269e493 gpopt::CXformSubqueryUnnest::Transform + 73
6    0x000064d61269e5f6 gpopt::CXformSubqueryUnnest::Transform + 322
7    0x000064d6125b00f0 gpopt::CGroupExpression::Transform + 948
8    0x000064d6125ce920 gpopt::CJobTransformation::EevtTransform + 220
9    0x000064d6125cef32 gpopt::CJobStateMachine + 376
10   0x000064d6125cea65 gpopt::CJobTransformation::FExecute + 123
11   0x000064d6125d5d63 gpopt::CScheduler::FExecute + 81
12   0x000064d6125d56aa gpopt::CScheduler::ExecuteJobs + 128
13   0x000064d6125d5622 gpopt::CScheduler::Run + 54
14   0x000064d6124d4b99 gpopt::CEngine::Optimize + 939
15   0x000064d61259bd0b gpopt::COptimizer::PexprOptimize + 115
16   0x000064d61259b62b gpopt::COptimizer::PdxlnOptimize + 1435
17   0x000064d612713a0d COptTasks::OptimizeTask + 1565
18   0x000064d612338f2b gpos::CTask::Execute + 199
19   0x000064d61233bab7 gpos::CWorker::Execute + 197
20   0x000064d6123372a7 gpos::CAutoTaskProxy::Execute + 227
21   0x000064d61233de4a gpos_exec + 933
22   0x000064d612712604 COptTasks::Execute + 226
23   0x000064d6127148d2 COptTasks::GPOPTOptimizedPlan + 178
24   0x000064d61271586a CGPOptimizer::GPOPTOptimizedPlan + 92
25   0x000064d612715f6d GPOPTOptimizedPlan + 39
26   0x000064d611f3abc1 optimize_query + 483
27   0x000064d611f1b1ce standard_planner + 168
28   0x000064d611f1b108 planner + 397
29   0x000064d612026d81 pg_plan_query + 142
30   0x000064d611d011ac <symbol not found> + 110870584627628
31   0x000064d611d00d47 ExplainQuery + 1431
                                                     QUERY PLAN                                                      
---------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=683.83..683.84 rows=1 width=0)
   ->  Insert on returning_parttab  (cost=683.83..683.84 rows=1 width=0)
         InitPlan 1 (returns $0)  (slice3)
           ->  Aggregate  (cost=683.81..683.82 rows=1 width=4)
                 ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=683.75..683.80 rows=1 width=4)
                       ->  Aggregate  (cost=683.75..683.76 rows=1 width=4)
                             ->  Append  (cost=0.00..567.00 rows=15567 width=4)
                                   ->  Seq Scan on returning_parttab_1_prt_1  (cost=0.00..567.00 rows=15567 width=4)
         ->  Result  (cost=0.00..0.01 rows=1 width=0)
 Optimizer: Postgres query optimizer
(10 rows)

```

a bit simplier query with subselect (instead of CTE) will also fail inside Orca:
```
postgres=# explain insert into returning_parttab values (1, 5, 'test') returning *, 1 + (select max(distkey) from returning_parttab);
INFO:  GPORCA failed to produce a plan, falling back to planner
DETAIL:  CNormalizer.cpp:1467: Failed assertion: FLocalColsSubsetOfInputCols(mp, pexprOutput)
Stack trace:
1    0x000064d612321b5f gpos::CErrorContext::Record + 191
2    0x000064d612322d0e gpos::CException::Raise + 278
3    0x000064d61254b1c4 gpopt::CNormalizer::PexprPullUpProjections + 270
4    0x000064d61269e414 gpopt::CXformSubqueryUnnest::PexprSubqueryUnnest + 1096
5    0x000064d61269e493 gpopt::CXformSubqueryUnnest::Transform + 73
6    0x000064d61269e5f6 gpopt::CXformSubqueryUnnest::Transform + 322
7    0x000064d6125b00f0 gpopt::CGroupExpression::Transform + 948
8    0x000064d6125ce920 gpopt::CJobTransformation::EevtTransform + 220
9    0x000064d6125cef32 gpopt::CJobStateMachine + 376
10   0x000064d6125cea65 gpopt::CJobTransformation::FExecute + 123
11   0x000064d6125d5d63 gpopt::CScheduler::FExecute + 81
12   0x000064d6125d56aa gpopt::CScheduler::ExecuteJobs + 128
13   0x000064d6125d5622 gpopt::CScheduler::Run + 54
14   0x000064d6124d4b99 gpopt::CEngine::Optimize + 939
15   0x000064d61259bd0b gpopt::COptimizer::PexprOptimize + 115
16   0x000064d61259b62b gpopt::COptimizer::PdxlnOptimize + 1435
17   0x000064d612713a0d COptTasks::OptimizeTask + 1565
18   0x000064d612338f2b gpos::CTask::Execute + 199
19   0x000064d61233bab7 gpos::CWorker::Execute + 197
20   0x000064d6123372a7 gpos::CAutoTaskProxy::Execute + 227
21   0x000064d61233de4a gpos_exec + 933
22   0x000064d612712604 COptTasks::Execute + 226
23   0x000064d6127148d2 COptTasks::GPOPTOptimizedPlan + 178
24   0x000064d61271586a CGPOptimizer::GPOPTOptimizedPlan + 92
25   0x000064d612715f6d GPOPTOptimizedPlan + 39
26   0x000064d611f3abc1 optimize_query + 483
27   0x000064d611f1b1ce standard_planner + 168
28   0x000064d611f1b108 planner + 397
29   0x000064d612026d81 pg_plan_query + 142
30   0x000064d611d011ac <symbol not found> + 110870584627628
31   0x000064d611d00d47 ExplainQuery + 1431
                                                     QUERY PLAN                                                      
---------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=683.82..683.83 rows=1 width=0)
   ->  Insert on returning_parttab  (cost=683.82..683.83 rows=1 width=0)
         InitPlan 1 (returns $0)  (slice3)
           ->  Aggregate  (cost=683.81..683.82 rows=1 width=4)
                 ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=683.75..683.80 rows=1 width=4)
                       ->  Aggregate  (cost=683.75..683.76 rows=1 width=4)
                             ->  Append  (cost=0.00..567.00 rows=15567 width=4)
                                   ->  Seq Scan on returning_parttab_1_prt_1  (cost=0.00..567.00 rows=15567 width=4)
         ->  Result  (cost=0.00..0.01 rows=1 width=0)
 Optimizer: Postgres query optimizer
(10 rows)

```

whitehawk commented on src/backend/gporca/libgpopt/src/operators/CLogicalReturning.cpp:
```diff
@@ -0,0 +1,204 @@
+//---------------------------------------------------------------------------
+//	Greenplum Database
+//	Copyright (C) 2013 EMC Corp.
```
I think the copyright headers for new files should be changed, I don't think we can use this particular one. But which one is correct is yet unclear to me, I'm checking it internally. So this item is blocked on my side.

kooooootb committed:
dont check logical dml that used columns are subset of input columns

kooooootb committed:
typo

kooooootb committed:
linter

kooooootb commented:
I reached the point where 
```
SET optimizer_trace_fallback=ON;

CREATE TABLE returning_parttab (distkey int4, partkey int4, t text)
DISTRIBUTED BY (distkey)
PARTITION BY RANGE (partkey) (START (1) END (10));

explain with cte as (select max(distkey) from returning_parttab) insert into returning_parttab values (1, 5, 'test') returning *, 1 + (select * from cte);
```

is now producing "no plan was computed for required properties" error
I found that this query works:
- if CTE is replaced by equivalent subquery
- if table is not partitioned
- if DML operation is replaced by select for returning expressions

whitehawk commented on src/backend/gporca/libgpopt/src/operators/CLogicalReturning.cpp:
```diff
@@ -0,0 +1,204 @@
+//---------------------------------------------------------------------------
+//	Greenplum Database
+//	Copyright (C) 2013 EMC Corp.
```
Please use follofing copyright header for now for all new src files:
```suggestion
//	Copyright (c) 2025 Greengage Community
```

kooooootb replied:
done

kooooootb committed:
copyright header

whitehawk commented:
Please wrap the description into 80 char width text, for ex:
```
Add support for RETURNING operator for ORCA

this commit adds RETURNING operator support for ORCA for DML operations

- in case of non-empty returning, use query's returningList to fill query's 
OutputColumns array and (if required) create in DXL query tree LogicalProject 
node that contains projection rules and Logical DML query node (LogicalInsert, 
etc);
- change PhysicalDML node's property computing methods to return all table's 
columns (except CTAS case where behaviour was unchanged), previously it 
outputted only changed columns;
- Add ProjList into generated DML plan's node which shows columns that DML node 
can return to parent node, in plan serialization to DXL stage it loses all 
unused columns for optimization;
- change nodeDML in executor to return modified tuples and set returning 
projections, similarly to ModifyTable node;
- update test samples in gporca and regress tests.
```

whitehawk commented:
> I reached the point where
> 
> ```
> SET optimizer_trace_fallback=ON;
> 
> CREATE TABLE returning_parttab (distkey int4, partkey int4, t text)
> DISTRIBUTED BY (distkey)
> PARTITION BY RANGE (partkey) (START (1) END (10));
> 
> explain with cte as (select max(distkey) from returning_parttab) insert into returning_parttab values (1, 5, 'test') returning *, 1 + (select * from cte);
> ```
> 
> is now producing "no plan was computed for required properties" error I found that this query works:
> 
> * if CTE is replaced by equivalent subquery
> * if table is not partitioned
> * if DML operation is replaced by select for returning expressions

Ok, it is Ok at this point. The edge case will be fixed later separately.

Please add both queries to the tests.

kooooootb committed:
fix subplans: add intermediate Result node if subplan-join's child node doesn't return enougn columns

problem occurs in case DML returns modified columns with replicated distribution and columns from subplan, which selects data from other table

Subplan inside ORCA is planned as NLJoin
On Expr2DXL stage subplan is pushed to projlist of nearest Result node above join node, join node itself is not needed so it is not added to DXL tree

Problem is that GatherMotion is added above DML with replicated distribution to return data to qdir
But this motion is under nearest Result node so when subplan is executed, its data has replicated distribution and there is no other motion to gather it

This commit adds intermediate Result node right before motion, after join, in case when join requires more columns than child node can derive
This will cause subplan to execute where is was meant to be executed while planning in ORCA

kooooootb committed:
linter

kooooootb committed:
fix logical error in triggers

kooooootb committed:
add test DML on partitioned table with RETURNING subquery with cte

kooooootb commented:
@whitehawk @andr-sokolov I added test for issue mentioned above as suggested
Also fixed regress and unit tests for dml with subqueries
pls review

kooooootb committed:
fix regress: add sort in test query

andr-sokolov commented on src/include/nodes/plannodes.h:
```diff
@@ -1330,6 +1330,7 @@ typedef struct DML
 	AttrNumber	tupleoidColIdx;	/* index of tuple oid column into the target list */
 	AttrNumber	tableoidColIdx; /* index of table oid column into the target list */
 	bool		canSetTag;		/* calculate processed tuples */
+	List	   *projTargetList;	/* target list for projection of outer node */
```
Why do we need to make changes in the core part, but not Orca only?

kooooootb replied:
ORCA operates DML plan nodes for dml operations, while planner uses ModifyTable nodes
So, we need to make DML nodes output results of modifying operations, like in ModifyTable

PR adds building projections for the returning tuples in DML plan node

DML also uses projections for the results of outer nodes (which are used to get values which will be e.g. inserted)
So, for DML plan nodes we need to store two lists of projections: for results of outer nodes and for returning tuples

There is common struct for all plans - Plan, which has field for list of projections
In ModifyTable this field is used to store returning projections and afaik ModifyTable uses results of outer nodes without projections
So, for DML plan node we need to store additional list of projections in its struct

kooooootb committed:
review fixes

andr-sokolov commented:
The ModifyTable structure contans the returningLists field which is a list of lists. Why is list enough for the DML node? Why not list of lists too?

kooooootb commented:
@andr-sokolov 
> The ModifyTable structure contans the returningLists field which is a list of lists. Why is list enough for the DML node? Why not list of lists too?

First of all i runned regress tests on planner (using optimizer=off, I checked that it used postgres planner in results) with commented out block of code
```
		if (parse->returningList)
			returningLists = lappend(returningLists,
									 subroot.parse->returningList);
```
which appends list to the returningLists field in ModifyTable in planner and **all tests passed**.

Then i runned these queries in PostgreSQL:
```
create table some_tab (f1 int, f2 int, f3 int, check (f1 < 10) no inherit);
create table some_tab_child (f4 int) inherits(some_tab);
explain update some_tab set f3 = 11 where f1 = 12 and f2 = 13 RETURNING *;
```
which uses block of code above
It prints plan
```
                                      QUERY PLAN                                       
---------------------------------------------------------------------------------------
 Update on some_tab  (cost=0.00..40.61 rows=1 width=14)
   Update on some_tab_child some_tab_1
   ->  Result  (cost=0.00..40.61 rows=1 width=14)
         ->  Seq Scan on some_tab_child some_tab_1  (cost=0.00..40.60 rows=1 width=10)
               Filter: ((f1 = 12) AND (f2 = 13))
(5 rows)
```

In Greenplum running UPDATE with optimizer_trace_fallback=on; results in
```
INFO:  GPORCA failed to produce a plan, falling back to planner
DETAIL:  Feature not supported: Inherited tables
                                   QUERY PLAN                                   
--------------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..2435.00 rows=5 width=20)
   ->  Update on some_tab  (cost=0.00..2435.00 rows=2 width=20)
         ->  Seq Scan on some_tab  (cost=0.00..1268.50 rows=1 width=18)
               Filter: ((f1 = 12) AND (f2 = 13))
         ->  Seq Scan on some_tab_child  (cost=0.00..1166.50 rows=1 width=22)
               Filter: ((f1 = 12) AND (f2 = 13))
 Optimizer: Postgres query optimizer
(7 rows)
```

Orca can't update base table, it throws exception with unsupported feature when I try to modify any base table.
And planner generates different plan with 1 DML node 

If we update child table (with no tables inheritted from it) Orca will generate plan
```
explain update some_tab_child set f4 = 11;
```
it prints
```
                                    QUERY PLAN                                     
-----------------------------------------------------------------------------------
 Update  (cost=0.00..431.08 rows=1 width=1)
   ->  Split  (cost=0.00..431.00 rows=1 width=30)
         ->  Result  (cost=0.00..431.00 rows=1 width=30)
               ->  Seq Scan on some_tab_child  (cost=0.00..431.00 rows=1 width=26)
 Optimizer: Pivotal Optimizer (GPORCA)
(5 rows)
```

So i think ModifyTable stores list of lists in the returningLists of ModifyTable for the case we modify base tables, to also modify tables inheritted from it, 1 list for 1 table.
Basically, in PostgreSQL 1 DML plan node can work with several relations and each needs returning list.
GPDB is currently unable to modify table, when it has other tables inheritted from it, so we currently don't need to store list of lists in DML node.

andr-sokolov commented on src/backend/gporca/libgpopt/src/operators/CLogicalReturning.cpp:
```diff
@@ -0,0 +1,203 @@
+{
+}
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CLogicalDML::CLogicalDML
+//
+//	@doc:
+//		Ctor
+//
+//---------------------------------------------------------------------------
```
The comment is about CLogicalDML::CLogicalDML, but the function name is CLogicalReturning::CLogicalReturning. The comment doesn't contain useful information. We can understand that the function is ctor by its name. I suggest to remove uninformative comments or make them informative

kooooootb replied:
I forgot about CLogicalDML
"Ctor" comment is placed there similarly to all other operators in this folder, i guess maybe doxygen parsers need those so it looks pretty in docs

kooooootb replied:
fixed CLogicalDML

andr-sokolov commented on src/backend/gporca/libgpopt/src/operators/CLogicalReturning.cpp:
```diff
@@ -0,0 +1,203 @@
+//---------------------------------------------------------------------------
+CLogicalReturning::CLogicalReturning(CMemoryPool *mp,
+									 CTableDescriptor *ptabdesc)
+	: CLogical(mp), m_ptabdesc(ptabdesc), m_pdrgpcrOutput(NULL)
```
Why do we need to set m_pdrgpcrOutput to NULL here? Is this value used some where? This value is reset in the 38th line.

kooooootb replied:
removed this

andr-sokolov commented on src/backend/executor/nodeDML.c:
```diff
@@ -348,9 +445,30 @@ ExecEndDML(DMLState *node)
+		result = ExecDML(node);
+		if (!result)
+			break;
+	}
```
We don't need the return value of  ExecDML, we should only check if the function returns NULL. What about
```suggestion
	while (ExecDML(node) != NULL);
```
to make the code simpler?

kooooootb replied:
done

andr-sokolov commented on src/backend/executor/nodeDML.c:
```diff
@@ -45,158 +65,199 @@ ExecDMLExplainEnd(PlanState *planstate, struct StringInfoData *buf)
+		{
+			isUpdate = true;
+		}
 
```
Unrelated change

kooooootb committed:
change field name to targetListProj

kooooootb committed:
reformulate test dedscription

kooooootb committed:
fix comment CLogicalDML in CLogicalReturning

kooooootb committed:
fix odd assignment

kooooootb committed:
shorten Squelch function in nodeDML

kooooootb committed:
add newline to nodeDML.c

andr-sokolov commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -4189,6 +4201,19 @@ CTranslatorDXLToPlStmt::TranslateDXLDml(
+		TranslateDXLProjList(project_list_output_dxlnode, &base_table_context,
+							 child_contexts, output_context);
+
+	m_has_returning = m_has_returning || plan->targetlist != NIL;
+
+	if (plan->targetlist != NIL &&
```
Is it possible that m_has_returning is true without target list? What about
```suggestion
	m_has_returning = (plan->targetlist != NIL);

	if (m_has_returning &&
```
?

kooooootb replied:
It can be possible if we have multiple dml nodes in plan
But currently GPDB only accepts plans with no more than one modifying node so this logic is for future, when GPDB will be able to handle multiple dml nodes

kooooootb replied:
done

andr-sokolov commented on src/backend/gporca/libgpopt/src/operators/CLogicalDML.cpp:
```diff
@@ -225,9 +225,11 @@ CLogicalDML::PopCopyWithRemappedColumns(CMemoryPool *mp,
+		mp, m_edmlop, m_ptabdesc, colref_array, pdrgpcrOutput, m_pbsModified,
+		pcrAction, pcrCtid, pcrSegmentId, pcrTupleOid, pcrTableOid);
+
+	return result;
```
What about adding `pdrgpcrOutput` only? The new variable (`result`) is redundant. Without it the patch will be smaller

kooooootb replied:
done

kooooootb committed:
smaller patch in clogicaldml.cpp

kooooootb committed:
unary or in dxl2plstmt

andr-sokolov commented on src/backend/gporca/libgpopt/src/operators/CLogicalInsert.cpp:
```diff
@@ -231,9 +255,9 @@ CLogicalInsert::OsPrint(IOstream &os) const
+	os << "], ";
 
-	return os;
+	return CLogicalReturning::OsPrint(os);
```
I run queries:
```
create table t(i int);
explain delete from t;
```

Running `p pexprLogDelete->DbgStr()` in GDB before the patch I get 
```
CLogicalDelete ("t"), Deleted Columns: ["i" (0)], "ctid" (1), "gp_segment_id" (7)
```
But after the patch:
```
CLogicalDelete ("t"), Deleted Columns: ["i" (0)], "ctid" (1), "gp_segment_id" (7), Output Columns: ["i" (8), "ctid" (9), "xmin" (10), "cmin" (11), "xmax" (12), "cmax" (13), "tableoid" (14), "gp_segment_id" (15)] Key sets: {[1,7]}
```

Why do I get "Output Columns" for the query without returing?

andr-sokolov commented on src/backend/gporca/libgpopt/src/operators/CLogicalInsert.cpp:
```diff
@@ -124,9 +155,15 @@ CLogicalInsert::PopCopyWithRemappedColumns(CMemoryPool *mp,
 {
 	CColRefArray *colref_array =
 		CUtils::PdrgpcrRemap(mp, m_pdrgpcrSource, colref_mapping, must_exist);
+
+	CColRefArray *pdrgpcrOutput =
+		CLogicalReturning::CopyRemappedColumns(mp, colref_mapping, must_exist);
 	m_ptabdesc->AddRef();
 
-	return GPOS_NEW(mp) CLogicalInsert(mp, m_ptabdesc, colref_array);
+	CLogicalInsert *result = GPOS_NEW(mp)
+		CLogicalInsert(mp, m_ptabdesc, colref_array, pdrgpcrOutput);
+
+	return result;
```
Let's simplify this change too

andr-sokolov commented on src/backend/executor/nodeDML.c:
```diff
@@ -206,12 +268,13 @@ DMLState*
 ExecInitDML(DML *node, EState *estate, int eflags)
 {
 	/* check for unsupported flags */
-	Assert(!(eflags & (EXEC_FLAG_BACKWARD | EXEC_FLAG_MARK | EXEC_FLAG_REWIND)));
```
Why is the EXEC_FLAG_REWIND flag supported now?

andr-sokolov commented on src/backend/gporca/libgpopt/src/operators/CLogicalDML.cpp:
```diff
@@ -225,9 +225,9 @@ CLogicalDML::PopCopyWithRemappedColumns(CMemoryPool *mp,
-	return GPOS_NEW(mp)
-		CLogicalDML(mp, m_edmlop, m_ptabdesc, colref_array, m_pbsModified,
-					pcrAction, pcrCtid, pcrSegmentId, pcrTupleOid, pcrTableOid);
+	return GPOS_NEW(mp) CLogicalDML(
+		mp, m_edmlop, m_ptabdesc, colref_array, pdrgpcrOutput, m_pbsModified,
```
A new code style should be as the code around. It is not typical to Orca that function name and the first function argument are placed in different lines

andr-sokolov commented on src/backend/gporca/libgpopt/src/operators/CLogicalReturning.cpp:
```diff
@@ -0,0 +1,203 @@
+			os << ", ";
+		}
+		os << "[";
+		ULONG ulPrintedKeys = 0;
```
What about BOOL instead of ULONG? I think we don't need count, we need a flag.

andr-sokolov commented on src/backend/gporca/libgpopt/src/operators/CLogicalReturning.cpp:
```diff
@@ -0,0 +1,203 @@
+									   UlongToColRefMap *colref_mapping,
+									   BOOL must_exist)
+{
+	CColRefArray *pdrgpcrOutput = NULL;
+	if (must_exist)
+	{
+		pdrgpcrOutput =
+			CUtils::PdrgpcrRemapAndCreate(mp, m_pdrgpcrOutput, colref_mapping);
+	}
+	else
+	{
+		pdrgpcrOutput = CUtils::PdrgpcrRemap(mp, m_pdrgpcrOutput,
+											 colref_mapping, must_exist);
+	}
+
+	return pdrgpcrOutput;
```
What about don't assing NULL and using ternary operator to emphasise that the return value of the both functions is assinged to one variable? I think that the pdrgpcrOutput is redundant and the code will be simpler without it
```suggestion
	return (must_exist) ? CUtils::PdrgpcrRemapAndCreate(mp, m_pdrgpcrOutput,
														colref_mapping)
						: CUtils::PdrgpcrRemap(mp, m_pdrgpcrOutput,
											   colref_mapping, must_exist);
```

andr-sokolov commented on src/test/regress/expected/partition_pruning_optimizer.out:
```diff
@@ -3492,15 +3492,14 @@ explain (verbose, costs off) delete from test_1_prt_extra where j = 2;
                                                      QUERY PLAN                                                      
 ---------------------------------------------------------------------------------------------------------------------
  Delete
-   Output: "outer".ColRef_0009, test_1_prt_extra.ctid
```
Why are this line and the same lines below not output after the patch?

kooooootb replied:
Before this patch DML plan nodes outputted all columns, which they get from child nodes to perform modification
Now they output only columns, required by parents (in this case it's parentless so no output)

andr-sokolov replied:
Can we don't change plans for queries without RETURNING?

andr-sokolov replied:
Let's create a new PR to rewrite using the targetlist field of the Plan structure

andr-sokolov commented on src/test/regress/expected/returning_gp_optimizer.out:
```diff
@@ -0,0 +1,231 @@
+       2 |      19 | multi2 3
+(11 rows)
+
+-- Test DML on partitioned table with RETURNING subquery with cte
+explain (costs off)
+with cte as (
+  select distkey from returning_parttab order by distkey limit 1
+) insert into returning_parttab values (1, 5, 'test') returning (select * from cte);
+INFO:  GPORCA failed to produce a plan, falling back to planner
+DETAIL:  No plan has been computed for required properties
```
Why do Orca fail to produce a plan? Why is this test case related to the patch?

kooooootb replied:
@whitehawk asked to add test for this case and said, somewhere at the end of March, that edge cases can be fixed later separately
I guess it has to do something with partition keys calculating in orca
